File: 2024/day2/part1.lua
dofile("UsefulStuff.lua")


local function part1(fileName)

	-- Loop over each report
	local validReports = 0
	for line in io.lines(Script_path()..fileName) do
		
		-- Parse report into an array
		local report = {}
		for value in string.gmatch(line,"%d+") do
			table.insert(report, tonumber(value) )
		end

		-- 
		local inc, dec = true,true
		for i = 2,#report do
			diff = report[i] - report[i-1]
			if math.abs(diff) > 3 or diff == 0 then
				inc,dec = false,false
				break
			elseif diff > 0 then
				dec = false
			else
				inc = false
			end
		end
		if inc or dec then
			validReports = validReports + 1
		end
	end

	-- Return the number of valid reports
	return validReports
end

UnitTest(part1, "testcases/test1.txt",2)

local result = part1("input.txt")
assert(result == 299, "Wrong answer")
print("Part 1: " .. tostring(result))

File: 2025/day9/Javascript/Part1.js
const { assert } = require('console');
const fs = require('fs');

function solution(filepath) {
    data = fs.readFileSync(filepath, 'UTF8').toString()
    .replaceAll(/\r\n/g,"\n").split('\n').map(x => x.split(",").map(y => Number(y)))

    maxSize = 0
    for (let i = 0; i < data.length; i++) {
        for (let j = i+1; j < data.length; j++) {
            maxSize = Math.max(maxSize,(Math.abs(data[i][0]-data[j][0])+1) * (Math.abs(data[i][1]-data[j][1])+1))
        }
    }
    return maxSize
}

assert(solution(__dirname + '/../testcases/test1.txt') == 50)
assert(solution(__dirname + '/../input.txt') == 4790063600)
console.log("Part 1:", solution(__dirname + '/../input.txt'))